docs/codeql/codeql-language-guides/using-api-graphs-in-ruby.rst

```diff
@@ -0,0 +1,183 @@
+that has not been extracted. Thus, it may be necessary to account for this in the models you write.
+For example, the ``ActionController::Base`` class has a predefined subclass ``Rails::ApplicationController``. To find
+all subclasses of ``ActionController::Base``, you must explicitly include the subclasses of ``Rails::ApplicationController`` as well.
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+
+    API::Node actionController() {
+      result =
+        [
+          API::getTopLevelMember("ActionController").getMember("Base"),
+          API::getTopLevelMember("Rails").getMember("ApplicationController")
+        ].getASubclass*()
+    }
+
+    select actionController()
+
+
+Using the API graph in dataflow queries
+---------------------------------------
+
+Dataflow queries often search for points where data from external sources enters the code base
+as well as places where data leaves the code base. API graphs provide a convenient way to refer 
+to external API components such as library functions and their inputs and outputs. API graph nodes
+cannot be used directly in dataflow queries they model entities that are defined externally,
+while dataflow nodes correspond to entities defined in the current code base. To brigde this gap
+the API node classes provide the ``asSource()`` and ``asSink()`` methods.
+
+The ``asSource()`` method is used to select dataflow nodes where a value from an external source
+enters the current code base. A typical example is the return value of a library function such as
+``File.read(path)``:
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+    select API::getTopLevelMember("File").getMethod("read").getParameter(1).asSource()
+
+
+The ``asSink()`` method is used to select dataflow nodes where a value leaves the
+current code base and flows into an external library. For example the second parameter
+of the ``File.write(path, value)`` method.
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+    select API::getTopLevelMember("File").getMethod("write").getParameter(1).asSink()
+
+A more complex example is a call to ``File.open`` with a block argument. This function creates a ```File`` instance
+and passes it to the supplied block. In this case the first parameter of the block is the place where an
+externally created value enters the code base, i.e. the ``|file|`` in the example below:
+
+.. code-block:: ruby
+
+    File.open("/my/file.txt", "w") { |file| file << "Hello world" }
+
+The following snippet finds parameters of blocks of ``File.open`` method calls:
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+    select API::getTopLevelMember("File").getMethod("open").getBlock().getParameter(0).asSource()
+
+The following example is a dataflow query that that uses API graphs to find cases where data that 
+is read flows into a call to ```File.write``.
+
+.. code-block:: ql
```

Comment:
Is there a reason why we used `File.open` above as an example and then use `File.read` here? Could we use `File.read` above so that the example follows through?

Reply:
`File.read` and `File.write` are simpler functions to read/write files. They were mentioned a little earlier in the text. The `File.open` method is more versatile and complex. I used it as an example to illustrate how to refer to  places where data flows into the parameter of a block (the `|file|` in the example). 

The final large example of a dataflow query is meant to show how API graphs and dataflow fit together in a more or less realistic query. I chose to use `File.read` and `File.write` because these are simpler sources/sinks . 

The `|file|` argument in the `File.open` example is not the file contents but a file handle from/to which we can read/write data. We could extend the example query with that, to make it even more realistic, however, it would also get a bit more complex as we need to track calls to `read` and `write` on the `|file|` parameter to identify additional sources and sinks for the query.

Reply:
Thanks for explaining. I'm guessing that your target audience will be aware of all of the above.